### PR TITLE
TextControl: Convert component to TypeScript

### PR DESCRIPTION
**Pull request by walbo**
## What?
Converts the `TextControl` component to TypeScript.

## Why?
Part of the @wordpress/components's TypeScript migration (https://github.com/WordPress/gutenberg/issues/35744).

## How?
Refactors the current `TextControl` to TypeScript.

## Testing Instructions
1. Run npm run dev and ensure no typing errors
2. Fire up the Storybook and confirm the `TextControl` continues to function as expected
3. Test the various text controls within the block/site editors in Gutenberg still work

cc: @ciampo

### CODE REVIEW

File: packages/components/tsconfig.json
Comment (ciampo) [inline packages/components/tsconfig.json:11]: Well spotted, it must be a leftover from my previous PR #39436

File: packages/components/src/text-control/types.ts
Comment (ciampo) [inline packages/components/src/text-control/types.ts]: We usually wrap props in `WordPressComponentProps` where the component is defined, would you mind following the same convention here too?

<details>
<summary>Example</summary>

```diff
diff --git a/packages/components/src/text-control/index.tsx b/packages/components/src/text-control/index.tsx
index 1ff2325f08..7634e751d1 100644
--- a/packages/components/src/text-control/index.tsx
+++ b/packages/components/src/text-control/index.tsx
@@ -13,10 +13,11 @@ import { forwardRef } from '@wordpress/element';
  * Internal dependencies
  */
 import BaseControl from '../base-control';
+import type { WordPressComponentProps } from '../ui/context';
 import type { TextControlProps } from './types';
 
 function UnforwardedTextControl(
-	props: TextControlProps,
+	props: WordPressComponentProps< TextControlProps, 'input', false >,
 	ref: ForwardedRef< HTMLInputElement >
 ) {
 	const {
diff --git a/packages/components/src/text-control/types.ts b/packages/components/src/text-control/types.ts
index 71f6d822c4..e72efdb611 100644
--- a/packages/components/src/text-control/types.ts
+++ b/packages/components/src/text-control/types.ts
@@ -7,28 +7,23 @@ import type { HTMLInputTypeAttribute } from 'react';
  * Internal dependencies
  */
 import type { BaseControlProps } from '../base-control/types';
-import type { WordPressComponentProps } from '../ui/context';
 
-export type TextControlProps = WordPressComponentProps<
-	Pick<
-		BaseControlProps,
-		'className' | 'hideLabelFromVision' | 'help' | 'label'
-	> & {
-		/**
-		 * A function that receives the value of the input.
-		 */
-		onChange: ( value: string ) => void;
-		/**
-		 * The current value of the input.
-		 */
-		value: string | number;
-		/**
-		 * Type of the input element to render. Defaults to "text".
-		 *
-		 * @default text
-		 */
-		type?: HTMLInputTypeAttribute;
-	},
-	'input',
-	false
->;
+export type TextControlProps = Pick<
+	BaseControlProps,
+	'className' | 'hideLabelFromVision' | 'help' | 'label'
+> & {
+	/**
+	 * A function that receives the value of the input.
+	 */
+	onChange: ( value: string ) => void;
+	/**
+	 * The current value of the input.
+	 */
+	value: string | number;
+	/**
+	 * Type of the input element to render. Defaults to "text".
+	 *
+	 * @default text
+	 */
+	type?: HTMLInputTypeAttribute;
+};
```
</details>

File: packages/components/src/text-control/types.ts
Comment (ciampo) [inline packages/components/src/text-control/types.ts]:
```suggestion
		 * @default 'text'
```

File: packages/components/src/text-control/stories/index.tsx
Comment (ciampo) [inline packages/components/src/text-control/stories/index.tsx]: I think we'd prefer to have a default story as _vanilla_ as possible. We could add another story that showcases using `help` and `label`. There's no need to specify `hideLabelFromVision` and `type`, since those are already the default prop values.

```suggestion
export const Default: ComponentStory<
	typeof TextControl
> = DefaultTemplate.bind( {} );
Default.args = {};

export const WithLabelAndHelpText: ComponentStory<
	typeof TextControl
> = DefaultTemplate.bind( {} );
WithLabelAndHelpText.args = {
	...Default.args,
	label: 'Label Text',
	help: 'Help text to explain the input.',
};
```

File: packages/components/src/text-control/stories/index.tsx
Comment (ciampo) [inline packages/components/src/text-control/stories/index.tsx:24]: @mirka, what's your stance here? At the moment the `value` control is disabled since the story controls `TextControl`'s value via internal state.

File: packages/components/src/text-control/types.ts
Comment (ciampo) [inline packages/components/src/text-control/types.ts]: If I were to write this component from scratch, I would type `value` as a `string`.

Since this component is _not_ experimental (and since the purpose of these TypeScript migrations is to introduce the least amount of changes possible), we may want to keep this type as it currently is, but consider updating it in a follow-up (see also this recent related [conversation](https://github.com/WordPress/gutenberg/pull/40535#issuecomment-1112049865))

**Review by ciampo — COMMENTED** [review]
Thank you for helping out on the `@wordpress/components` TypeScript refactor, @walbo !

I just left a couple of inline comments, but this PR is in a great shape already.

Finally, would you mind adding an entry to the pachake's [CHANGELOG](https://github.com/WordPress/gutenberg/blob/1b58a2bfdc6b908c6712c9334365f45bd51ba15d/packages/components/CHANGELOG.md) ? You may want to add an `### Internal` section under `## Unreleased`.

File: packages/components/src/text-control/types.ts
Comment (walbo) [reply packages/components/src/text-control/types.ts]: Will update the PR. Thanks for the heads up.

File: packages/components/src/text-control/types.ts
Comment (walbo) [reply packages/components/src/text-control/types.ts]: I agree it should be `string` in this case. The old jsdocs said `string` but the README file says `string | number`.

File: packages/components/src/text-control/types.ts
Comment (walbo) [reply packages/components/src/text-control/types.ts]: Thanks.

File: packages/components/src/text-control/stories/index.tsx
Comment (walbo) [reply packages/components/src/text-control/stories/index.tsx]: Sounds good to me.

File: packages/components/src/text-control/index.tsx
Comment (mirka) [inline packages/components/src/text-control/index.tsx:19]: By the way, sorry about ef3a12e! I fixed up the contributing guide in #40682.

File: packages/components/src/text-control/stories/index.tsx
Comment (mirka) [reply packages/components/src/text-control/stories/index.tsx:24]: Sure, I think it's a reasonable pattern to follow for controlled stories.

File: packages/components/src/text-control/stories/index.tsx
Comment (mirka) [inline packages/components/src/text-control/stories/index.tsx]: Nit: It looks like we could remove the `control: { type: null }` in this case. Storybook hides the controls for `on*` props by default.

File: packages/components/src/text-control/stories/index.tsx
Comment (mirka) [inline packages/components/src/text-control/stories/index.tsx]: Apparently, the Action thing (enabled on [this line](https://github.com/WordPress/gutenberg/pull/40633/files#diff-f162be886476a0f6aaa0dc9958d453e89d1e56ce78a01b1918873dc958d51edeR21)) only works by passing through the `onChange` prop.

```diff
const DefaultTemplate: ComponentStory< typeof TextControl > = ( {
+	onChange,
	...args
} ) => {
	const [ value, setValue ] = useState( '' );

	return (
		<TextControl
			{ ...args }
			value={ value }
			onChange={ ( v ) => {
				setValue( v );
+				onChange( v );
```

https://user-images.githubusercontent.com/555336/165803385-b897a8e2-3ad4-44a8-95f6-091672672c9e.mp4

**Review by mirka — COMMENTED** [review]
Wonderful, thank you for the contribution!

Looks great on my end, I think we're good to go once the Action thing is sorted 👍

File: packages/components/src/text-control/types.ts
Comment (ciampo) [reply packages/components/src/text-control/types.ts]: Let's keep it as `string|number` for this PR. It would be great if we could then open a follow-up PR to change it to be only `string` — we could then assess in that PR if effectively it's a change that can be applied without introducing any breaking change.

Would you be up for that?

Comment (walbo) [reply packages/components/src/text-control/types.ts]: Yes. Lets do that. I will creeate a follow-up PR once this is merged.

File: packages/components/src/text-control/index.tsx
Comment (walbo) [reply packages/components/src/text-control/index.tsx:19]: No worries :) Great that you updated the guide. Did follow the guide and seems to cover most of the stuff you need.

File: packages/components/src/text-control/stories/index.tsx
Comment (walbo) [reply packages/components/src/text-control/stories/index.tsx]: Thanks. Added in [e30a8dd](https://github.com/WordPress/gutenberg/pull/40633/commits/e30a8ddf2e7abf6548bf3b71d0d4df7ddf66f280)

File: packages/components/src/text-control/stories/index.tsx
Comment (walbo) [reply packages/components/src/text-control/stories/index.tsx]: Thanks. I didn't know that 👍 Removed in [e30a8dd](https://github.com/WordPress/gutenberg/pull/40633/commits/e30a8ddf2e7abf6548bf3b71d0d4df7ddf66f280)

File: packages/components/src/text-control/index.tsx
Comment (mirka) [reply packages/components/src/text-control/index.tsx:19]: Oooh, good to hear! 🙌

**Review by mirka — APPROVED** [review]
Perfect, thank you! 🚀

File: packages/components/src/text-control/index.tsx
Comment (ciampo) [reply packages/components/src/text-control/index.tsx:19]: Thank you for the feedback! And of course do let us know if you feel like there's something missing (or something wrong) in the guide 😄

**Pull request merged**
